provider/cmd/main.go

```diff
@@ -4,18 +4,45 @@
 package main
 
 import (
+	"errors"
 	"fmt"
 	"os"
+	"syscall"
 
 	"github.com/ironcore-dev/libvirt-provider/provider/cmd/app"
+	originzap "go.uber.org/zap"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+const (
+	errTerminatedApp = "app terminated with error"
+)
+
 func main() {
 	ctx := ctrl.SetupSignalHandler()
 
-	if err := app.Command().ExecuteContext(ctx); err != nil {
-		fmt.Println(err.Error())
-		os.Exit(1)
+	cmd, ptrLogger := app.Command()
+	exitCode := int(0)
+	if err := cmd.ExecuteContext(ctx); err != nil {
+		exitCode = 1
+
+		// logger can be nil, if flags weren't parsed
+		if *ptrLogger != nil {
+			(*ptrLogger).Error(errTerminatedApp, originzap.NamedError("error", err))
+		} else {
+			fmt.Fprintln(os.Stderr, errTerminatedApp+": "+err.Error())
```

Comment:
Would it be possible to do something like `setupLog` here?
It's already defined in the `Run` method, should it be possible to bring it out at main, and pass it as an arg to Run then?

Reply:
ptrLogger contains pointer to setupLog which is created later.
But if you want to simplify code, we can specified what we expect from setupLogger.
Current setupLog is created from arguments (--zap-devel, --zap-encoder, ...) which can modified format of logs. If i create setupLog in main function, i will have problem because arguments are parsed in Execute function, so it can generate different format of log records as another loggers in program. Main problem here is RunE function which return error into main and it creates chicken-egg problem. We can switch to Run without error and we will print errors directly in Run function and in this case we can simplify code.

Reply:
if you want, i can prepare second solution for comparison in different branch

Reply:
@hardikdr  i checked the code and i think ideally solution will be stop use setup logger and use unspecified logger as default and for other components we can use named loggers. Main reason is wrong usage of setup logger and named logger in Run function, their are mixed somehow randomly. Result is setup logger or using of logger name for filtration of logger messages useless. we can discuss it on sync meeting but currently it looks like you don't have any rules what and when use what creates more questions as benefits.

Reply:
Thanks, I will check that one, and let's hold on to this one till we conclude.